pkg/instancetype/revision/store.go

```diff
@@ -43,33 +44,68 @@ import (
 )
 
 func (h *revisionHandler) Store(vm *virtv1.VirtualMachine) error {
-	instancetypeRevision, err := h.storeInstancetypeRevision(vm)
+	instancetypeStatusRef, err := h.storeInstancetypeRevision(vm)
 	if err != nil {
 		log.Log.Object(vm).Reason(err).Error("Failed to store ControllerRevision of VirtualMachineInstancetypeSpec for the Virtualmachine.")
 		return err
 	}
 
-	preferenceRevision, err := h.storePreferenceRevision(vm)
+	preferenceStatusRef, err := h.storePreferenceRevision(vm)
 	if err != nil {
 		log.Log.Object(vm).Reason(err).Error("Failed to store ControllerRevision of VirtualMachinePreferenceSpec for the Virtualmachine.")
 		return err
 	}
 
-	return h.patchVM(instancetypeRevision, preferenceRevision, vm)
+	return h.patchVM(instancetypeStatusRef, preferenceStatusRef, vm)
 }
 
-func (h *revisionHandler) storeInstancetypeRevision(vm *virtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
-	if vm.Spec.Instancetype == nil || vm.Spec.Instancetype.RevisionName != "" {
+func (h *revisionHandler) storeInstancetypeRevision(vm *virtv1.VirtualMachine) (*virtv1.InstancetypeStatusRef, error) {
+	if vm.Spec.Instancetype == nil {
 		return nil, nil
 	}
 
-	storedRevision, err := h.createInstancetypeRevision(vm)
-	if err != nil {
-		return nil, err
+	if vm.Status.InstancetypeRef == nil {
+		vm.Status.InstancetypeRef = &virtv1.InstancetypeStatusRef{}
+	}
+
+	matcher := vm.Spec.Instancetype
+	statusRef := vm.Status.InstancetypeRef.DeepCopy()
+
+	if matcher.Name != statusRef.Name {
+		statusRef.Name = matcher.Name
+		statusRef.ControllerRevisionRef = nil
+	}
+
+	if matcher.Kind != statusRef.Kind {
+		statusRef.Kind = matcher.Kind
+		statusRef.ControllerRevisionRef = nil
+	}
+
+	if matcher.RevisionName != "" {
+		if statusRef.ControllerRevisionRef == nil || statusRef.ControllerRevisionRef.Name != matcher.RevisionName {
+			statusRef.ControllerRevisionRef = &virtv1.ControllerRevisionRef{
+				Name: matcher.RevisionName,
+			}
+		}
+	}
+
+	if statusRef.ControllerRevisionRef == nil {
+		storedRevision, err := h.createInstancetypeRevision(vm)
+		if err != nil {
+			return nil, err
+		}
+
+		statusRef.ControllerRevisionRef = &virtv1.ControllerRevisionRef{
+			Name: storedRevision.Name,
+		}
+	}
+
+	if !equality.Semantic.DeepEqual(vm.Status.InstancetypeRef, statusRef) {
+		vm.Status.InstancetypeRef = statusRef
+		return vm.Status.InstancetypeRef, nil
 	}
 
-	vm.Spec.Instancetype.RevisionName = storedRevision.Name
-	return storedRevision, nil
+	return nil, nil
```

Comment:
Why return nil in some cases?

Reply:
It allows us to build a single patchset to use with `PatchStatus` when there are changes to either the instance type or preference references.